lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -725,17 +784,169 @@ struct GrandCentralPass : public GrandCentralBase<GrandCentralPass> {
+/// else get the path from the `nonlocalPath`. Emit error and return None, if
+/// there are multiple paths between `lcaModule` and `targetModule`. The caller
+/// must handle the None return value and error out.
+static Optional<ArrayRef<InstanceOp>>
+getPath(FModuleLike targetModule, FModuleOp lcaModule,
+        SmallVector<InstanceOp> &nonlocalPath, ApplyState &state) {
+  ArrayRef<InstanceOp> pathToSend;
+  if (targetModule == lcaModule)
+    return pathToSend;
+  auto isInstanceInlcaModule = [&](InstanceOp inst) {
+    return (inst->getParentOfType<FModuleOp>() == lcaModule);
+  };
+  if (nonlocalPath.empty()) {
+    // For all paths from root to `targetModule`. Get the path that passes
+    // through the `lcaModule`. Only one such path must exist.
+    for (auto path : state.instancePathcache.getAbsolutePaths(targetModule)) {
+      // If `lcaModule` exists in the path
+      if (llvm::find_if(path, isInstanceInlcaModule) != path.end()) {
+        // If already found one path, and another path exists, emit error.
+        if (!pathToSend.empty()) {
+          targetModule.emitOpError(
+              "cannot handle multiple paths from the parent module ")
+              << lcaModule.moduleNameAttr()
+              << " till the GrandCentral view target";
+          return None;
+        }
+
+        // Store the path from `lcaModule`.
+        pathToSend = path.drop_until(isInstanceInlcaModule);
+      }
+    }
+  } else // Get the path from the `lcaModule` in the `nonlocalPath`.
+    pathToSend = ArrayRef<InstanceOp>(
+        llvm::find_if(nonlocalPath, isInstanceInlcaModule), nonlocalPath.end());
+  return pathToSend;
+}
+
+/// Set the `pathFromSrcToCompanion` with the path from `srcTarget` to
+/// `companionTarget` via the `lcaModule`. The `lcaModule` denotes the least
+/// common ancestor for the source and destination modules. The first
+/// `InstanceOp` in the `pathFromSrcToCompanion` must instantiate the module of
+/// `srcTarget` and the last must instantiate the module `companionTarget`.
+static LogicalResult
+computeRefPortsPathViaLCA(AnnoPathValue &srcTarget,
+                          AnnoPathValue &companionTarget,
+                          SmallVector<InstanceOp> &pathFromSrcToCompanion,
+                          FModuleOp &lcaModule, ApplyState &state) {
+  if (isa<FExtModuleOp>(srcTarget.ref.getOp()))
+    return srcTarget.ref.getOp()->emitError(
+        "RefType source cannot be an extern module port, it must be moved to "
+        "the instance of the extern module");
+  pathFromSrcToCompanion.clear();
+
+  auto srcModule = cast<FModuleOp>(srcTarget.ref.getModule());
+  auto companionModule = cast<FModuleOp>(companionTarget.ref.getModule());
+  if (srcModule == companionModule)
+    return success();
+  auto pathToSrc = getPath(srcModule, lcaModule, srcTarget.instances, state);
+  if (!pathToSrc)
+    return mlir::emitError(lcaModule.getLoc(),
```

Comment:
Consider:
```suggestion
    return lcaModule.emitError(lcaModule,
```
(and few lines below, if go this way)